### PR TITLE
Fix CUDA error not getting captured by handler

### LINKED ISSUE #91758: C10_CUDA_KERNEL_LAUNCH_CHECK() becomes a no-op
### 🐛 Describe the bug

`C10_CUDA_KERNEL_LAUNCH_CHECK` calls `cudaGetLastError`:

https://github.com/pytorch/pytorch/blob/18b37bbff91d5746e5a1ec63bf82ce24a717434f/c10/cuda/CUDAException.h#L73

however, the result is discarded!

https://github.com/pytorch/pytorch/blob/18b37bbff91d5746e5a1ec63bf82ce24a717434f/c10/cuda/CUDAException.h#L27-L32

The code claims that the error will be obtained in `c10_cuda_check_implementation`, but that's not always true: many errors are non-sticky, meaning that after `cudaGetLastError` the error state is reset. In that case, this macro becomes a no-op.

In particular, launching a kernel with too many resources (too many registers & grid/block size) produces an error that's silently ignored by this macro -- and the kernel is silently not run :cry:  . Would be good to create a unittest about catching such errors.

It seems this bug was introduced in https://github.com/pytorch/pytorch/pull/85256 cc @ezyang @gchanan @zou3519 @ngimel @r-barnes .

### Versions

n/a

### PR DESCRIPTION
**Pull request by r-barnes**
Fixes #91758. Still leaves functions on the hotpath.

### CODE REVIEW

File: c10/cuda/CUDADeviceAssertionHost.cpp [hunk 1]
```diff
@@ -13,8 +13,17 @@
 #include <string>
 #include <thread>
 
-#define CHECK_CUDA_API_CALL_WITHOUT_CHECKING_DEVICE_ASSERTS() \
-  c10_cuda_check_implementation(__FILE__, __FUNCTION__, __LINE__, false)
+#define C10_CUDA_CHECK(EXPR)                                        \
```
Comment (ngimel) [inline c10/cuda/CUDADeviceAssertionHost.cpp:16]:
```suggestion
#define C10_CUDA_CHECK_WO_DSA(EXPR)                                        \
```

File: c10/cuda/CUDAException.h [hunk 2]
```diff
@@ -24,17 +24,16 @@ class C10_CUDA_API CUDAError : public c10::Error {
...
-    c10::cuda::c10_cuda_check_implementation(                            \
-        __FILE__,                                                        \
-        __func__, /* Line number's data type is not well-defined between \
-                      compilers, so we perform an explicit cast */       \
-        static_cast<uint32_t>(__LINE__),                                 \
-        true);                                                           \
+#define C10_CUDA_CHECK(EXPR)                                        \
+  do {                                                              \
+    const cudaError_t __err = EXPR;                                 \
+    c10::cuda::c10_cuda_check_implementation(                       \
```
Comment (ngimel) [inline c10/cuda/CUDAException.h:30]: you also need to change declaration on line 94

**Review by ngimel — APPROVED** [review]

**Review by malfet — APPROVED** [review]

**Review by malfet — COMMENTED** [review]
@r-barnes as the follow up, can we please add a test (using c++ extensions) to avoid those regressions in the future

**Pull request closed without merging**